Add convert_to_pil_image to pil util

robmarkcole opened the pull request:
## Proposed change
Adds a helper function to convert a valid input (usually an `image` returned by the image_processing integration) into a PIL Image. It appears partially corrupted images will cause an exception as this issue https://github.com/home-assistant/home-assistant/issues/31610, which this function will handle.

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/home-assistant/issues/31610
- Link to documentation pull request: 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

robmarkcole committed:
Update pil.py

robmarkcole commented:
Hmm I dont understand why we are getting `homeassistant/util/pil.py:9:0: E0611: No name 'UnidentifiedImageError' in module 'PIL' (no-name-in-module)`

balloob commented:
What version of PIL was it introduced vs what is installed with Home Assistant ?

robmarkcole commented:
OK it was introduced in 7.0 (latest) and we are currently on `pillow==6.2.1`

Release notes: Pillow will now throw a custom `UnidentifiedImageError` when an image cannot be identified. For backwards compatibility, this will inherit from `IOError`. 

Do we want to bump to 7.0 or just catch IOError?

balloob commented:
You can do a PR to bump PIL but you'll need to test it and make sure all the things work with it.

pvizeli commented:
Where is that function used?

robmarkcole commented:
it can be used by doods/tensorflow and also in Sighthound in future

robmarkcole commented:
it can be used by doods/tensorflow and also in Sighthound in future

balloob commented:
We should only add new functions if it's used.

robmarkcole committed:
Update doods and tensorflow to use convert_to_pil_image

robmarkcole commented:
Ok updated doods/tensorflow to use the function

robmarkcole committed:
Update pil.py

robmarkcole commented on homeassistant/components/doods/image_processing.py:
```diff
@@ -231,7 +230,9 @@ def device_state_attributes(self):
 
     def _save_image(self, image, matches, paths):
-        img = Image.open(io.BytesIO(bytearray(image))).convert("RGB")
+        img = convert_to_pil_image(image)
```
We could just pass the PIL image to _save_image actually

robmarkcole replied:
Actually this doesn't matter as we have already validated the image data in `process_image`

robmarkcole replied:
I think pil_image would be a better variable name..?

pvizeli reviewed (commented):
Do we need log a warning?

balloob commented on homeassistant/util/pil.py:
```diff
@@ -2,9 +2,18 @@
+    """Convert a camera image to a PIL Image."""
+    try:
+        return Image.open(io.BytesIO(bytearray(image))).convert("RGB")
+    except UnidentifiedImageError:
```
One thing that is weird here, is that now we are not giving any feedback to the user. Something just didn't work. However, this function cannot know what to report, as it doesn't know if it's an upload or how to refer to the image in a way that the user understands. 

I don't think that having this function is beneficial as we're promoting bad practice of swallowing errors.

robmarkcole replied:
Perhaps it would be better to implement this functionality as a `validate_image_data` method on the `Image_processing` class, and log the error? I think this is better than having uncaught exceptions as this issue https://github.com/home-assistant/home-assistant/issues/31610

balloob replied:
I agree. Let's have Tensorflow catch it and report on it.

robmarkcole replied:
OK added

balloob replied:
I'm honestly not sure if this function is _really_ helping at all. It's saving 1 line and it's less clear what is going on ? I think that we should just put this `try…except` in doods and TF.

pvizeli replied:
I think we should trow our own exception and use `try...except` instead the `if not image:`.

robmarkcole replied:
It is not saving much code at this time but it will standardise the way PIL is used, and avoid the same issues coming up time and again if used. But I am open to guidance on the best way forward. Re throwing an HA exception, is there a suitable one, and do we not prefer logs to exceptions?

balloob replied:
I think that the PIL exception is fine, we should just call PIL directly in those integrations and don't have a helper function.

robmarkcole replied:
To clarify, we want to catch the UnidentifiedImageError exception and log it, or are we going to allow it to just surface with no try.. except..?

balloob replied:
we are going to catch it inside `doods` and `tensorflow`. Copy paste the content of this function and paste it into `doods/__init__.py` etc.

robmarkcole committed:
Add log messages on bad data

robmarkcole committed:
Drop convert_to_pil_image

Just perform conversion in the integrations without seperate convert_to_pil_image()

robmarkcole commented:
OK dropped the special function as discussed
